File: app/src/main/java/com/example/myapplication/protocol/PacketStatsTracker.kt
package com.example.myapplication.protocol

data class PacketStatsSnapshot(
    val packetsReceived: Long,
    val packetsLost: Long,
)

data class PacketRecordResult(
    val snapshot: PacketStatsSnapshot,
    val gapCount: Long = 0L,
    val expectedCounter: Long? = null,
    val actualCounter: Long? = null,
)

class PacketStatsTracker {
    private var lastCounter = UNINITIALIZED_COUNTER
    private var packetsReceived = 0L
    private var packetsLost = 0L

    fun reset() {
        lastCounter = UNINITIALIZED_COUNTER
        packetsReceived = 0L
        packetsLost = 0L
    }

    fun record(counter: Long): PacketRecordResult {
        var gapCount = 0L
        var expectedCounter: Long? = null
        if (lastCounter != UNINITIALIZED_COUNTER) {
            expectedCounter = lastCounter + 1
            if (counter > expectedCounter) {
                gapCount = counter - expectedCounter
                packetsLost += gapCount
            }
        }

        packetsReceived++
        lastCounter = counter

        return PacketRecordResult(
            snapshot = snapshot(),
            gapCount = gapCount,
            expectedCounter = expectedCounter,
            actualCounter = counter,
        )
    }

    fun snapshot(): PacketStatsSnapshot {
        return PacketStatsSnapshot(
            packetsReceived = packetsReceived,
            packetsLost = packetsLost,
        )
    }

    private companion object {
        const val UNINITIALIZED_COUNTER = -1L
    }
}

File: ios/Woona/Protocol/PacketStatsTracker.swift
import Foundation

struct PacketStatsSnapshot: Equatable {
    let packetsReceived: UInt64
    let packetsLost: UInt64
}

struct PacketRecordResult: Equatable {
    let snapshot: PacketStatsSnapshot
    let gapCount: UInt64
    let expectedCounter: UInt64?
    let actualCounter: UInt64?
}

final class PacketStatsTracker {
    private var lastCounter: UInt64?
    private var packetsReceived: UInt64 = 0
    private var packetsLost: UInt64 = 0

    func reset() {
        lastCounter = nil
        packetsReceived = 0
        packetsLost = 0
    }

    func record(counter: UInt64) -> PacketRecordResult {
        var gapCount: UInt64 = 0
        let expectedCounter = lastCounter.map { $0 + 1 }

        if let expectedCounter, counter > expectedCounter {
            gapCount = counter - expectedCounter
            packetsLost += gapCount
        }

        packetsReceived += 1
        lastCounter = counter

        return PacketRecordResult(
            snapshot: snapshot(),
            gapCount: gapCount,
            expectedCounter: expectedCounter,
            actualCounter: counter
        )
    }

    func snapshot() -> PacketStatsSnapshot {
        PacketStatsSnapshot(packetsReceived: packetsReceived, packetsLost: packetsLost)
    }
}
